src/block/decompress_safe.rs

```diff
@@ -322,19 +287,33 @@ fn duplicate_overlapping_slice(
         return Err(DecompressError::OffsetOutOfBounds);
     }
     if offset == 1 {
-        let val = sink.as_slice()[start];
-        let dest = sink.pos();
-        sink.output[dest..dest + match_length].fill(val);
-        sink.pos += match_length;
+        let val = sink.filled_slice()[start];
+        sink.extend_with_fill(val, match_length);
     } else {
-        for i in start..start + match_length {
-            let b = sink.as_slice()[i];
-            sink.push(b);
-        }
+        sink.extend_from_within_overlapping(start, match_length);
     }
     Ok(())
 }
 
+/// Decompress all bytes of `input` into `output`.
+/// `output` should be preallocated with a size of of the uncompressed data.
+#[inline]
+pub fn decompress_into(input: &[u8], output: &mut [u8]) -> Result<usize, DecompressError> {
```

Comment:
`output_pos` has to be removed as it's unsound. Any user can trivially provide `output[output_pos..]` themselves.

Sink `pos` has other effects on the decoder other than being an "output offset". It allows the decoder to "look back" on the bytes at `output[..output_pos]`.